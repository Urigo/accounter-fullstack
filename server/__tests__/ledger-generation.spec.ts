import { Application, testkit } from 'graphql-modules';
import 'reflect-metadata';
import { beforeAll, describe, it } from 'vitest';
import { fetchChargeQuery, fetchChargesQuery, getDummyApp } from './ledger-test-data';

let app: Application;

describe.concurrent('Ledger Generation', () => {
  beforeAll(async () => {
    app = getDummyApp();
  });

  describe.concurrent('Local currency (ILS)', () => {
    it('should generate ledger records for basic ILS charge', async ({ expect }) => {
      const result = await testkit.execute(app, {
        document: fetchChargeQuery,
        variableValues: {
          id: '1000',
        },
      });

      expect(result.data?.charge?.ledgerRecords.message).toBeUndefined();

      expect(result.data?.charge?.ledgerRecords?.records?.length).toBe(2);

      expect(result.data?.charge?.ledgerRecords).toMatchSnapshot();
    });

    it('dates diff should have no effect', async ({ expect }) => {
      const result = await testkit.execute(app, {
        document: fetchChargeQuery,
        variableValues: {
          id: '1001',
        },
      });

      expect(result.data?.charge?.ledgerRecords.message).toBeUndefined();

      expect(result.data?.charge?.ledgerRecords?.records?.length).toBe(2);

      expect(result.data?.charge?.ledgerRecords).toMatchSnapshot();
    });

    it('Should handle no-invoice businesses', async ({ expect }) => {
      const result = await testkit.execute(app, {
        document: fetchChargeQuery,
        variableValues: {
          id: '1002',
        },
      });

      expect(result.data?.charge?.ledgerRecords.message).toBeUndefined();

      expect(result.data?.charge?.ledgerRecords?.records?.length).toBe(1);

      expect(result.data?.charge?.ledgerRecords).toMatchSnapshot();
    });
  });

  describe.concurrent('Foreign currency', () => {
    it('should generate ledger records for basic USD charge', async ({ expect }) => {
      const result = await testkit.execute(app, {
        document: fetchChargeQuery,
        variableValues: {
          id: '1100',
        },
      });
      expect(result.data?.charge?.ledgerRecords.message).toBeUndefined();
      expect(result.data?.charge?.ledgerRecords?.records?.length).toBe(2);
      expect(result.data?.charge?.ledgerRecords).toMatchSnapshot();
    });

    it('should generate ledger records including conversion diff ledger for USD charge with dates diff', async ({
      expect,
    }) => {
      const result = await testkit.execute(app, {
        document: fetchChargeQuery,
        variableValues: {
          id: '1101',
        },
      });
      expect(result.data?.charge?.ledgerRecords.message).toBeUndefined();
      expect(result.data?.charge?.ledgerRecords?.records?.length).toBe(3);
      expect(result.data?.charge?.ledgerRecords).toMatchSnapshot();
      const exchangeLedger = result.data?.charge?.ledgerRecords?.records?.[2];
      expect(exchangeLedger?.description).toBe('Exchange ledger record');
    });

    it('Conversion diff ledger should be generated based on transaction debit_date', async ({
      expect,
    }) => {
      const result = await testkit.execute(app, {
        document: fetchChargesQuery,
        variableValues: {
          IDs: ['1102', '1103'],
        },
      });
      expect(result.data?.chargesById?.length).toBe(2);

      // charge 1102: only debit_daet diff
      const chargeDebitDateDiff = result.data?.chargesById?.[0];
      expect(chargeDebitDateDiff?.ledgerRecords?.records?.length).toBe(3);

      // charge 1103: only event_date diff
      const chargeEventDateDiff = result.data?.chargesById?.[1];
      expect(chargeEventDateDiff?.ledgerRecords?.records?.length).toBe(2);
    });

    it('Conversion diff ledger should should have opposite sides based on what payment/doc chronological order', async ({
      expect,
    }) => {
      const result = await testkit.execute(app, {
        document: fetchChargesQuery,
        variableValues: {
          IDs: ['1101', '1104'],
        },
      });
      expect(result.data?.chargesById?.length).toBe(2);

      // charge 1101: doc first
      const chargeDocFirst = result.data?.chargesById?.[0];
      expect(chargeDocFirst?.ledgerRecords?.records?.length).toBe(3);
      const exchangeLedger1 = chargeDocFirst?.ledgerRecords?.records?.[2];
      expect(exchangeLedger1?.debitAccount1?.id).toBe('900');

      // charge 1104: transaction first
      const chargeTransactionFirst = result.data?.chargesById?.[1];
      expect(chargeTransactionFirst?.ledgerRecords?.records?.length).toBe(3);
      const exchangeLedger2 = chargeDocFirst?.ledgerRecords?.records?.[2];
      expect(exchangeLedger2?.creditAccount1?.id).toBe('2003');
    });

    it('Should handle no-invoice businesses', async ({ expect }) => {
      const result = await testkit.execute(app, {
        document: fetchChargeQuery,
        variableValues: {
          id: '1105',
        },
      });

      expect(result.data?.charge?.ledgerRecords.message).toBeUndefined();

      expect(result.data?.charge?.ledgerRecords?.records?.length).toBe(1);

      expect(result.data?.charge?.ledgerRecords).toMatchSnapshot();
    });
  });

  describe.concurrent('Conversion', () => {
    it('should generate ledger records for basic conversion charge', async ({ expect }) => {
      const result = await testkit.execute(app, {
        document: fetchChargeQuery,
        variableValues: {
          id: '1200',
        },
      });
      expect(result.data?.charge?.ledgerRecords.message).toBeUndefined();
      expect(result.data?.charge?.ledgerRecords?.records?.length).toBe(2);
      expect(result.data?.charge?.ledgerRecords).toMatchSnapshot();
    });

    it('Unbalances charge should throw', async ({ expect }) => {
      const result = await testkit.execute(app, {
        document: fetchChargeQuery,
        variableValues: {
          id: '1201',
        },
      });
      expect(result.data?.charge?.ledgerRecords.message).toContain(
        'Conversion charges must have a ledger balance',
      );
    });

    it('should generate ledger records for conversion charge with dates diff', async ({
      expect,
    }) => {
      const result = await testkit.execute(app, {
        document: fetchChargeQuery,
        variableValues: {
          id: '1202',
        },
      });
      expect(result.data?.charge?.ledgerRecords.message).toBeUndefined();
      expect(result.data?.charge?.ledgerRecords?.records?.length).toBe(2);
      expect(result.data?.charge?.ledgerRecords).toMatchSnapshot();
    });
  });

  describe.concurrent('Misc', () => {
    it('should handle currency AND date diff', async ({ expect }) => {
      // the odd case of Lance's charge
      const result = await testkit.execute(app, {
        document: fetchChargeQuery,
        variableValues: {
          id: '1900',
        },
      });
      expect(result.data?.charge?.ledgerRecords.message).toBeUndefined();
      expect(result.data?.charge?.ledgerRecords?.records?.length).toBe(3);
      expect(result.data?.charge?.ledgerRecords).toMatchSnapshot();
    });
  });
});
