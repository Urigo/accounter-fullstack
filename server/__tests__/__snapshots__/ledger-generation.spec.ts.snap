// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Ledger Generation > Conversion > should generate ledger records for basic conversion charge 1`] = `
{
  "records": [
    {
      "creditAccount1": {
        "name": "Foreign Business 2",
      },
      "creditAccount2": null,
      "creditAmount1": {
        "formatted": "€ 100,000.00",
      },
      "creditAmount2": null,
      "debitAccount1": {
        "name": "Account ID:...1",
      },
      "debitAccount2": null,
      "debitAmount1": {
        "formatted": "€ 100,000.00",
      },
      "debitAmount2": null,
      "description": "Description for transaction 4200a",
      "invoiceDate": 2020-01-01T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 200,000.00",
      },
      "localCurrencyCreditAmount2": null,
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 200,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": "1",
      "valueDate": 2020-01-01T00:00:00.000Z,
    },
    {
      "creditAccount1": {
        "name": "Account ID:...1",
      },
      "creditAccount2": null,
      "creditAmount1": null,
      "creditAmount2": null,
      "debitAccount1": {
        "name": "Foreign Business 1",
      },
      "debitAccount2": null,
      "debitAmount1": null,
      "debitAmount2": null,
      "description": "Description for transaction 4200b",
      "invoiceDate": 2020-01-01T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 200,000.00",
      },
      "localCurrencyCreditAmount2": null,
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 200,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": "1",
      "valueDate": 2020-01-01T00:00:00.000Z,
    },
  ],
}
`;

exports[`Ledger Generation > Conversion > should generate ledger records for conversion charge with dates diff 1`] = `
{
  "records": [
    {
      "creditAccount1": {
        "name": "Foreign Business 2",
      },
      "creditAccount2": null,
      "creditAmount1": {
        "formatted": "€ 100,000.00",
      },
      "creditAmount2": null,
      "debitAccount1": {
        "name": "Account ID:...1",
      },
      "debitAccount2": null,
      "debitAmount1": {
        "formatted": "€ 100,000.00",
      },
      "debitAmount2": null,
      "description": "Description for transaction 4202a",
      "invoiceDate": 2020-01-02T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 250,000.00",
      },
      "localCurrencyCreditAmount2": null,
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 250,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": "1",
      "valueDate": 2020-01-02T00:00:00.000Z,
    },
    {
      "creditAccount1": {
        "name": "Account ID:...1",
      },
      "creditAccount2": null,
      "creditAmount1": null,
      "creditAmount2": null,
      "debitAccount1": {
        "name": "Foreign Business 1",
      },
      "debitAccount2": null,
      "debitAmount1": null,
      "debitAmount2": null,
      "description": "Description for transaction 4202b",
      "invoiceDate": 2020-01-01T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 250,000.00",
      },
      "localCurrencyCreditAmount2": null,
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 250,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": "1",
      "valueDate": 2020-01-01T00:00:00.000Z,
    },
  ],
}
`;

exports[`Ledger Generation > Foreign currency > Should handle no-invoice businesses 1`] = `
{
  "records": [
    {
      "creditAccount1": {
        "name": "Account ID:...1",
      },
      "creditAccount2": null,
      "creditAmount1": {
        "formatted": "$ 100,000.00",
      },
      "creditAmount2": null,
      "debitAccount1": {
        "name": "Foreign Business 2",
      },
      "debitAccount2": null,
      "debitAmount1": {
        "formatted": "$ 100,000.00",
      },
      "debitAmount2": null,
      "description": "Description for transaction 4105",
      "invoiceDate": 2020-01-01T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 100,000.00",
      },
      "localCurrencyCreditAmount2": null,
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 100,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": "2",
      "valueDate": 2020-01-01T00:00:00.000Z,
    },
  ],
}
`;

exports[`Ledger Generation > Foreign currency > should generate ledger records for basic USD charge 1`] = `
{
  "records": [
    {
      "creditAccount1": {
        "name": "Foreign Business 1",
      },
      "creditAccount2": null,
      "creditAmount1": {
        "formatted": "$ 100,000.00",
      },
      "creditAmount2": null,
      "debitAccount1": {
        "name": "Tax Category 3",
      },
      "debitAccount2": null,
      "debitAmount1": {
        "formatted": "$ 100,000.00",
      },
      "debitAmount2": null,
      "description": null,
      "invoiceDate": 2020-01-01T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 100,000.00",
      },
      "localCurrencyCreditAmount2": null,
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 100,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": null,
      "valueDate": 2020-01-01T00:00:00.000Z,
    },
    {
      "creditAccount1": {
        "name": "Account ID:...1",
      },
      "creditAccount2": null,
      "creditAmount1": {
        "formatted": "$ 100,000.00",
      },
      "creditAmount2": null,
      "debitAccount1": {
        "name": "Foreign Business 1",
      },
      "debitAccount2": null,
      "debitAmount1": {
        "formatted": "$ 100,000.00",
      },
      "debitAmount2": null,
      "description": "Description for transaction 4100",
      "invoiceDate": 2020-01-01T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 100,000.00",
      },
      "localCurrencyCreditAmount2": null,
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 100,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": "1",
      "valueDate": 2020-01-01T00:00:00.000Z,
    },
  ],
}
`;

exports[`Ledger Generation > Foreign currency > should generate ledger records including conversion diff ledger for USD charge with dates diff 1`] = `
{
  "records": [
    {
      "creditAccount1": {
        "name": "Foreign Business 1",
      },
      "creditAccount2": null,
      "creditAmount1": {
        "formatted": "$ 100,000.00",
      },
      "creditAmount2": null,
      "debitAccount1": {
        "name": "Tax Category 3",
      },
      "debitAccount2": null,
      "debitAmount1": {
        "formatted": "$ 100,000.00",
      },
      "debitAmount2": null,
      "description": null,
      "invoiceDate": 2020-01-01T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 100,000.00",
      },
      "localCurrencyCreditAmount2": null,
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 100,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": null,
      "valueDate": 2020-01-01T00:00:00.000Z,
    },
    {
      "creditAccount1": {
        "name": "Account ID:...1",
      },
      "creditAccount2": null,
      "creditAmount1": {
        "formatted": "$ 100,000.00",
      },
      "creditAmount2": null,
      "debitAccount1": {
        "name": "Foreign Business 1",
      },
      "debitAccount2": null,
      "debitAmount1": {
        "formatted": "$ 100,000.00",
      },
      "debitAmount2": null,
      "description": "Description for transaction 4100",
      "invoiceDate": 2020-01-02T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 150,000.00",
      },
      "localCurrencyCreditAmount2": null,
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 150,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": "1",
      "valueDate": 2020-01-02T00:00:00.000Z,
    },
    {
      "creditAccount1": {
        "name": "Foreign Business 1",
      },
      "creditAccount2": null,
      "creditAmount1": null,
      "creditAmount2": null,
      "debitAccount1": {
        "name": "Exchange",
      },
      "debitAccount2": null,
      "debitAmount1": null,
      "debitAmount2": null,
      "description": "Exchange ledger record",
      "invoiceDate": 2020-01-02T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 50,000.00",
      },
      "localCurrencyCreditAmount2": null,
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 50,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": "1",
      "valueDate": 2020-01-02T00:00:00.000Z,
    },
  ],
}
`;

exports[`Ledger Generation > Local currency (ILS) > Should handle no-invoice businesses 1`] = `
{
  "records": [
    {
      "creditAccount1": {
        "name": "Account ID:...1",
      },
      "creditAccount2": null,
      "creditAmount1": null,
      "creditAmount2": null,
      "debitAccount1": {
        "name": "Local Business 2 (no invoices)",
      },
      "debitAccount2": null,
      "debitAmount1": null,
      "debitAmount2": null,
      "description": "Description for transaction 4002",
      "invoiceDate": 2020-01-01T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 117,000.00",
      },
      "localCurrencyCreditAmount2": null,
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 117,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": "1",
      "valueDate": 2020-01-01T00:00:00.000Z,
    },
  ],
}
`;

exports[`Ledger Generation > Misc > should handle currency AND date diff 1`] = `
{
  "records": [
    {
      "creditAccount1": {
        "name": "Tax Category 1",
      },
      "creditAccount2": {
        "name": "VAT [temp]",
      },
      "creditAmount1": {
        "formatted": "$ 83,000.00",
      },
      "creditAmount2": {
        "formatted": "$ 17,000.00",
      },
      "debitAccount1": {
        "name": "Local Business 1",
      },
      "debitAccount2": null,
      "debitAmount1": {
        "formatted": "$ 100,000.00",
      },
      "debitAmount2": null,
      "description": null,
      "invoiceDate": 2020-01-01T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 83,000.00",
      },
      "localCurrencyCreditAmount2": {
        "formatted": "₪ 17,000.00",
      },
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 100,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": null,
      "valueDate": 2020-01-01T00:00:00.000Z,
    },
    {
      "creditAccount1": {
        "name": "Account ID:...1",
      },
      "creditAccount2": null,
      "creditAmount1": null,
      "creditAmount2": null,
      "debitAccount1": {
        "name": "Local Business 1",
      },
      "debitAccount2": null,
      "debitAmount1": null,
      "debitAmount2": null,
      "description": "Description for transaction 4900",
      "invoiceDate": 2020-01-02T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 117,000.00",
      },
      "localCurrencyCreditAmount2": null,
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 117,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": "1",
      "valueDate": 2020-01-02T00:00:00.000Z,
    },
    {
      "creditAccount1": {
        "name": "Local Business 1",
      },
      "creditAccount2": null,
      "creditAmount1": null,
      "creditAmount2": null,
      "debitAccount1": {
        "name": "Exchange",
      },
      "debitAccount2": null,
      "debitAmount1": null,
      "debitAmount2": null,
      "description": "Exchange ledger record",
      "invoiceDate": 2020-01-02T00:00:00.000Z,
      "localCurrencyCreditAmount1": {
        "formatted": "₪ 217,000.00",
      },
      "localCurrencyCreditAmount2": null,
      "localCurrencyDebitAmount1": {
        "formatted": "₪ 217,000.00",
      },
      "localCurrencyDebitAmount2": null,
      "reference1": "1",
      "valueDate": 2020-01-02T00:00:00.000Z,
    },
  ],
}
`;
